packages/utils/src/backoff.ts

```diff
@@ -0,0 +1,42 @@
+import { durations } from './time'
+
+export async function wait(milliseconds: number): Promise<void> {
+  return new Promise((resolve) => setTimeout(resolve, milliseconds))
+}
+
+export async function backoff(
+  fn: () => Promise<any>,
+  options: {
+    minDelay?: number
+    maxDelay?: number
+    delayMultiple?: number
+  } = {}
+): ReturnType<typeof fn> {
+  const { minDelay = durations.seconds(1), maxDelay = durations.minutes(10), delayMultiple = 2 } = options
+  let delay: number | undefined
+
+  if (minDelay >= maxDelay) throw Error('minDelay should be smaller than maxDelay')
+  else if (delayMultiple < 1) throw Error('delayMultiple should be larger than 1')
+
+  return new Promise<ReturnType<typeof fn>>(async (resolve, reject) => {
+    const tick = async () => {
+      try {
+        const result = await fn()
+        return resolve(result)
+      } catch (err) {
+        if (delay === maxDelay) {
+          return reject(err)
+        }
+
+        // if delay is not set, our first delay is minDelay
+        // else we start exp increasing
+        delay = !delay ? minDelay : Math.min(delay * delayMultiple, maxDelay)
+
+        await wait(delay)
+        return tick()
```

Comment:
```suggestion
```

Reply:
I want the function to be able to resolve to `fn`'s return, removing this will not let that happen

Reply:
I think so, because the value `result` is passed into `Promise.resolve()` - the return value of the promise constructor is anyway ignored

Reply:
simplified the code, re-open this if its still an issue